www/js/diary/timelineHelper.ts

```diff
@@ -212,3 +226,345 @@ const locations2GeojsonTrajectory = (trip, locationList, trajectoryColor?) => {
+ * after the end of the trip, so the trip end transition of a processed
+ * trip may actually show up as an unprocessed transition.
+ * See https://github.com/e-mission/e-mission-phone/issues/214#issuecomment-286279163
+ *
+ * Let's abstract this out into our own minor state machine.
+ */
+const transitions2Trips = function (transitionList: Array<ServerData<TripTransition>>) {
+  let inTrip = false;
+  const tripList = [];
+  let currStartTransitionIndex = -1;
+  let currEndTransitionIndex = -1;
+  let processedUntil = 0;
+
+  while (processedUntil < transitionList.length) {
+    // Logger.log("searching within list = "+JSON.stringify(transitionList.slice(processedUntil)));
+    if (inTrip == false) {
+      const foundStartTransitionIndex = transitionList
+        .slice(processedUntil)
+        .findIndex(isStartingTransition);
+      if (foundStartTransitionIndex == -1) {
+        logDebug('No further unprocessed trips started, exiting loop');
+        processedUntil = transitionList.length;
+      } else {
+        currStartTransitionIndex = processedUntil + foundStartTransitionIndex;
+        processedUntil = currStartTransitionIndex;
+        logDebug(
+          'Unprocessed trip started at ' + JSON.stringify(transitionList[currStartTransitionIndex]),
+        );
+        inTrip = true;
+      }
+    } else {
+      const foundEndTransitionIndex = transitionList
+        .slice(processedUntil)
+        .findIndex(isEndingTransition);
+      if (foundEndTransitionIndex == -1) {
+        logDebug(
+          "Can't find end for trip starting at " +
+            JSON.stringify(transitionList[currStartTransitionIndex]) +
+            ' dropping it',
+        );
+        processedUntil = transitionList.length;
+      } else {
+        currEndTransitionIndex = processedUntil + foundEndTransitionIndex;
+        processedUntil = currEndTransitionIndex;
+        logDebug(`currEndTransitionIndex ${currEndTransitionIndex}`);
+        logDebug(
+          'Unprocessed trip starting at ' +
+            JSON.stringify(transitionList[currStartTransitionIndex]) +
+            ' ends at ' +
+            JSON.stringify(transitionList[currEndTransitionIndex]),
+        );
+        tripList.push([
+          transitionList[currStartTransitionIndex],
+          transitionList[currEndTransitionIndex],
+        ]);
+        inTrip = false;
+      }
+    }
+  }
+  return tripList;
+};
+
+const linkTrips = function (trip1, trip2) {
+  // complete trip1
+  trip1.starting_trip = { $oid: trip2.id };
+  trip1.exit_fmt_time = trip2.enter_fmt_time;
+  trip1.exit_local_dt = trip2.enter_local_dt;
+  trip1.exit_ts = trip2.enter_ts;
+
+  // start trip2
+  trip2.ending_trip = { $oid: trip1.id };
+  trip2.enter_fmt_time = trip1.exit_fmt_time;
+  trip2.enter_local_dt = trip1.exit_local_dt;
+  trip2.enter_ts = trip1.exit_ts;
+};
+
+export const readUnprocessedTrips = function (startTs, endTs, lastProcessedTrip) {
```

Comment:
Ditto. I think that most of the functions that have been copied over do not have a type defined